plugins/libplugin-pay.c

```diff
@@ -660,129 +661,298 @@ static void payment_result_infer(struct route_hop *route,
-	p->result = tal_sendpay_result_from_json(p, buffer, toks);
+	const struct node_id *id = &p->route[report_pos].nodeid;
 
-	if (p->result == NULL) {
+	if (report_pos == 0) {
 		plugin_log(p->plugin, LOG_UNUSUAL,
-			   "Unable to parse `waitsendpay` result: %.*s",
-			   json_tok_full_len(toks),
-			   json_tok_full(buffer, toks));
-		payment_set_step(p, PAYMENT_STEP_FAILED);
-		payment_continue(p);
-		return command_still_pending(cmd);
+			   "Node #%zu (%s) claimed we sent them invalid %s",
+			   report_pos + 1,
+			   type_to_string(tmpctx, struct node_id, id),
+			   style);
+	} else {
+		plugin_log(p->plugin, LOG_UNUSUAL,
+			   "Node #%zu (%s) claimed #%zu (%s) sent them invalid %s",
+			   report_pos + 1,
+			   type_to_string(tmpctx, struct node_id, id),
+			   report_pos,
+			   type_to_string(tmpctx, struct node_id,
+					  &p->route[report_pos-1].nodeid),
+			   style);
 	}
+}
 
-	payment_result_infer(p->route, p->result);
+static struct command_result *
+handle_final_failure(struct command *cmd,
+		     struct payment *p,
+		     const struct node_id *final_id,
+		     enum onion_type failcode)
+{
+	/* We use an exhaustive switch statement here so you get a compile
+	 * warning when new ones are added, and can think about where they go */
+	switch (failcode) {
+	case WIRE_FINAL_INCORRECT_CLTV_EXPIRY:
+		report_tampering(p, tal_count(p->route)-1, "cltv");
+		goto error;
+	case WIRE_FINAL_INCORRECT_HTLC_AMOUNT:
+		report_tampering(p, tal_count(p->route)-1, "amount");
+		goto error;
 
-	if (p->result->state == PAYMENT_COMPLETE) {
-		payment_set_step(p, PAYMENT_STEP_SUCCESS);
-		payment_continue(p);
-		return command_still_pending(cmd);
+	/* BOLT #4:
+	 *
+	 * A _forwarding node_ MAY, but a _final node_ MUST NOT:
+	 *...
+	 *     - return an `invalid_onion_version` error.
+	 *...
+	 *     - return an `invalid_onion_hmac` error.
+	 *...
+	 *     - return an `invalid_onion_key` error.
+	 *...
+	 *     - return a `temporary_channel_failure` error.
+	 *...
+	 *     - return a `permanent_channel_failure` error.
+	 *...
+	 *     - return a `required_channel_feature_missing` error.
+	 *...
+	 *     - return an `unknown_next_peer` error.
+	 *...
+	 *     - return an `amount_below_minimum` error.
+	 *...
+	 *     - return a `fee_insufficient` error.
+	 *...
+	 *     - return an `incorrect_cltv_expiry` error.
+	 *...
+	 *     - return an `expiry_too_soon` error.
+	 *...
+	 *     - return an `expiry_too_far` error.
+	 *...
+	 *     - return a `channel_disabled` error.
+	 */
+	case WIRE_INVALID_ONION_VERSION:
+	case WIRE_INVALID_ONION_HMAC:
+	case WIRE_INVALID_ONION_KEY:
+	case WIRE_TEMPORARY_CHANNEL_FAILURE:
+	case WIRE_PERMANENT_CHANNEL_FAILURE:
+	case WIRE_REQUIRED_CHANNEL_FEATURE_MISSING:
+	case WIRE_UNKNOWN_NEXT_PEER:
+	case WIRE_AMOUNT_BELOW_MINIMUM:
+	case WIRE_FEE_INSUFFICIENT:
+	case WIRE_INCORRECT_CLTV_EXPIRY:
+	case WIRE_EXPIRY_TOO_FAR:
+	case WIRE_EXPIRY_TOO_SOON:
+	case WIRE_CHANNEL_DISABLED:
+		goto strange_error;
```

Comment:
Since this failure code is from remote, the remote might suddenly give us some unknown failure code.  Of course, that means we just fall-through into `strange_error`, which is appropriate.